### ISSUE TITLE
Add eigenvalue functions and solvers for all BC/grid-type combinations

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/spectraldiffx#38]
## Summary

Add the missing eigenvalue functions and functional solvers (Layer 0 + Layer 1 module classes) for all boundary condition × grid type combinations. This covers **staggered Dirichlet** (DST-II/III), **regular Neumann** (DCT-I), and **mixed-BC** eigenvalues (DST-IV, DCT-IV, DST-III, DCT-III), plus the corresponding Helmholtz/Poisson solvers in **1D, 2D, and 3D**.

## Motivation

Both the [SailFFish paper](https://arxiv.org/abs/2301.01145) and [PoisFFT paper](https://arxiv.org/abs/1409.8116) document a complete matrix of BC type × grid type → transform type mappings. spectraldiffx already has all 4 DST and 4 DCT transform types implemented, but only uses 3 of the possible eigenvalue/solver combinations:

| BC | Grid | Transform | Eigenvalue | Status |
|---|---|---|---|---|
| Dirichlet | Regular (vertex) | DST-I | `dst1_eigenvalues` | ✅ Exists |
| Dirichlet | Staggered (cell) | DST-II/III | `dst2_eigenvalues` | ❌ Missing |
| Neumann | Regular (vertex) | DCT-I | `dct1_eigenvalues` | ❌ Missing |
| Neumann | Staggered (cell) | DCT-II/III | `dct2_eigenvalues` | ✅ Exists |
| Periodic | Either | FFT | `fft_eigenvalues` | ✅ Exists |
| Mixed: Dir+Neu (stag) | Staggered | DST-IV | `dst4_eigenvalues` | ❌ Missing |
| Mixed: Neu+Dir (stag) | Staggered | DCT-IV | `dct4_eigenvalues` | ❌ Missing |
| Mixed: Dir+Neu (reg) | Regular | DST-III/II | `dst3_eigenvalues` | ❌ Missing |
| Mixed: Neu+Dir (reg) | Regular | DCT-III/II | `dct3_eigenvalues` | ❌ Missing |

The staggered Dirichlet solver is critical for incompressible flow solvers (projection method) where the pressure Poisson equation lives on a cell-centered grid. The regular Neumann solver (DCT-I) is needed for vertex-centered Neumann problems. These are prerequisites for finitevolx's Arakawa C-grid integration.

### Naming: Explicit DST/DCT Type Numbers

Use DST/DCT type numbers in all function names — this is unambiguous and directly maps to the transform type:

| Function | Transform | BC + Grid |
|---|---|---|
| `solve_helmholtz_dst1` | DST-I | Dirichlet, regular |
| `solve_helmholtz_dst2` | DST-II/III | Dirichlet, staggered |
| `solve_helmholtz_dct1` | DCT-I | Neumann, regular |
| `solve_helmholtz_dct2` | DCT-II/III | Neumann, staggered |

Keep the old names (`solve_helmholtz_dst`, `solve_helmholtz_dct`) as **aliases** for backwards compatibility:
- `solve_helmholtz_dst` → alias for `solve_helmholtz_dst1`
- `solve_helmholtz_dct` → alias for `solve_helmholtz_dct2`

Same convention for eigenvalue functions, Poisson convenience wrappers, and module classes.

### Part A: Eigenvalue Functions (`eigenvalues.py`)

Add 6 new eigenvalue functions, each a one-liner following the established pattern:

**Staggered Dirichlet (DST-II basis):**
```python
def dst2_eigenvalues(N: int, dx: float) -> Float[Array, " N"]:
    """Dirichlet BCs on staggered (cell-centered) grid.
    λ_k = −4/dx² · sin²(π(k+1) / (2N)),   k = 0, …, N−1
    All eigenvalues strictly negative. Pairs with DST-II fwd / DST-III bwd.
    """
    k = jnp.arange(N)
    return -4.0 / dx**2 * jnp.sin(jnp.pi * (k + 1) / (2 * N)) ** 2
```

**Regular Neumann (DCT-I basis):**
```python
def dct1_eigenvalues(N: int, dx: float) -> Float[Array, " N"]:
    """Neumann BCs on regular (vertex-centered) grid.
    λ_k = −4/dx² · sin²(πk / (2(N−1))),   k = 0, …, N−1
    λ_0 = 0 (constant null mode). Requires N ≥ 2.
    Pairs with DCT-I fwd / DCT-I bwd (self-inverse).
    """
    k = jnp.arange(N)
    return -4.0 / dx**2 * jnp.sin(jnp.pi * k / (2 * (N - 1))) ** 2
```

**Mixed-BC eigenvalues** (all share formula `sin²(π(k+0.5)/(2N))`):
```python
def dst4_eigenvalues(N, dx):  # DirichletStag+NeumannStag, DST-IV (self-inverse)
def dct4_eigenvalues(N, dx):  # NeumannStag+DirichletStag, DCT-IV (self-inverse)
def dst3_eigenvalues(N, dx):  # Dirichlet+Neumann (regular), DST-III/II
def dct3_eigenvalues(N, dx):  # Neumann+Dirichlet (regular), DCT-III/II
```

These are numerically identical but semantically distinct (different BCs, different transforms). Can share a private helper internally.

### Part B: Functional Solvers (`solvers.py`)

**New 2D solvers:**

| Function | BC | Grid | Transforms | Eigenvalues |
|---|---|---|---|---|
| `solve_helmholtz_dst2` | Dirichlet | Staggered | DST-II fwd, DST-III bwd | `dst2_eigenvalues` |
| `solve_poisson_dst2` | Dirichlet | Staggered | (convenience wrapper) | |
| `solve_helmholtz_dct1` | Neumann | Regular | DCT-I fwd, DCT-I bwd | `dct1_eigenvalues` |
| `solve_poisson_dct1` | Neumann | Regular | (convenience wrapper) | |

For `solve_helmholtz_dst2`: `idstn(type=2)` automatically uses DST-III as inverse (SciPy convention), all eigenvalues strictly negative (no null-mode handling needed).

For `solve_helmholtz_dct1`: k=0 eigenvalue is zero (null mode), requires same zero-mean gauge as existing DCT-II solver.

**New 1D solvers** (following `solve_helmholtz_fft_1d` pattern):

| Function | BC | Grid |
|---|---|---|
| `solve_helmholtz_dst1_1d` | Dirichlet | Regular |
| `solve_helmholtz_dst2_1d` | Dirichlet | Staggered |
| `solve_helmholtz_dct2_1d` | Neumann | Staggered |
| `solve_helmholtz_dct1_1d` | Neumann | Regular |

Plus `solve_poisson_*` convenience wrappers for each.

**New 3D solvers** (extending the 2D pattern with a third axis):

| Function | BC | Grid |
|---|---|---|
| `solve_helmholtz_dst1_3d` | Dirichlet | Regular |
| `solve_helmholtz_dst2_3d` | Dirichlet | Staggered |
| `solve_helmholtz_dct2_3d` | Neumann | Staggered |
| `solve_helmholtz_dct1_3d` | Neumann | Regular |
| `solve_helmholtz_fft_3d` | Periodic | Either |

Plus `solve_poisson_*` convenience wrappers for each.

The 3D solvers follow the same pattern as 2D — apply the transform along 3 axes, build a 3D eigenvalue tensor `eigz[:, None, None] + eigy[None, :, None] + eigx[None, None, :]`, divide, inverse transform. Both SailFFish and PoisFFT validate this approach in 3D.

**Backwards-compatible aliases:**
```python
solve_helmholtz_dst = solve_helmholtz_dst1       # Dirichlet regular 2D
solve_poisson_dst = solve_poisson_dst1
solve_helmholtz_dct = solve_helmholtz_dct2       # Neumann staggered 2D  
solve_poisson_dct = solve_poisson_dct2
```

### Part C: Module Classes (`solvers.py`)

```python
class StaggeredDirichletHelmholtzSolver2D(eqx.Module):
    """2D solver with Dirichlet BCs on staggered grid (DST-II/III)."""
    dx: float
    dy: float
    alpha: float = 0.0
    def __call__(self, rhs): return solve_helmholtz_dst2(rhs, self.dx, self.dy, self.alpha)

class RegularNeumannHelmholtzSolver2D(eqx.Module):
    """2D solver with Neumann BCs on regular grid (DCT-I)."""
    dx: float
    dy: float
    alpha: float = 0.0
    def __call__(self, rhs): return solve_helmholtz_dct1(rhs, self.dx, self.dy, self.alpha)
```

### Part D: Public API Exports

Export all new functions and classes from:
- `spectraldiffx/_src/fourier/__init__.py`
- `spectraldiffx/__init__.py`

### Part E: Tests

**Eigenvalue tests** (`test_fourier_eigenvalues.py`) — for each new function:
- Shape: returns array of length N
- Sign: all strictly negative (Dirichlet) or has null mode at k=0 (Neumann regular)
- Known values: spot-check for small N against manual calculation
- dx scaling: eigenvalues scale as 1/dx²
- Cross-validation: `dst2_eigenvalues` differs from `dst1_eigenvalues` for same N

**Solver tests** (`test_fourier_solvers.py`) — for each new solver (1D, 2D, and 3D):
- Eigenfunction recovery: construct `rhs = eigenvalue × eigenfunction`, solve, compare
- Zero-mean gauge: for Neumann solvers, `mean(solution) ≈ 0`
- Zero RHS: should return zero
- Helmholtz vs Poisson: different λ produces different result

Eigenfunction test patterns:

| Transform | Eigenfunction |
|---|---|
| DST-I (Dirichlet regular) | `sin(π(k+1)(j+1)/(N+1))` |
| DST-II (Dirichlet staggered) | `sin(π(k+1)(2j+1)/(2N))` |
| DCT-I (Neumann regular) | `cos(πk·j/(N−1))` |
| DCT-II (Neumann staggered) | `cos(πk(2j+1)/(2N))` |

**Module class tests**: callable, produces correct output, works with `eqx.filter_jit`.

**3D solver tests**: same eigenfunction recovery pattern extended to 3D (product of 1D eigenfunctions along each axis).

## References

- SailFFish Table 1 (DFT Type and Eigenvalues for Bounded Solver Types)
- PoisFFT Tables 2–4 (Transforms, PS eigenvalues, FD2 eigenvalues)
- Schumann & Sweet (1988), "Fast Fourier transforms for direct solution of Poisson's equation with staggered boundary conditions"
- Full plan: `spectraldiffx/.plans/staggered-grid-solvers.md`
